alf/environments/suite_tic_tac_toe.py

```diff
@@ -0,0 +1,192 @@
+            env_id=self._env_ids,
+            env_info={
+                "play0_win": torch.zeros(self._batch_size),
+                "play1_win": torch.zeros(self._batch_size),
+                "draw": torch.zeros(self._batch_size),
+                "invalid_move": torch.zeros(self._batch_size),
+            })
+
+    def _step(self, action):
+        prev_game_over = self._game_over
+        prev_action = action.clone()
+        prev_action[prev_game_over] = 0
+        self._boards[prev_game_over] = self._empty_board
+        step_type = torch.full((self._batch_size, ), int(StepType.MID))
+        player = self._get_current_player().to(torch.float32)
+        x = action % 3
+        y = action // 3
+        valid = self._boards[self._B, y, x] == 0
+        self._boards[self._B[valid], y[valid], x[valid]] = player[valid]
+        won = self._check_player_win(player)
+        reward = torch.where(won, -player, torch.tensor(0.))
+        reward = torch.where(valid, reward, player)
+        game_over = self._check_game_over()
+        game_over = torch.max(game_over, ~valid)
+        step_type[game_over] = int(StepType.LAST)
+        step_type[prev_game_over] = int(StepType.FIRST)
+        discount = torch.ones(self._batch_size)
+        discount[game_over] = 0.
+        self._boards[prev_game_over] = self._empty_board
+        self._game_over = game_over
+        self._prev_action = action
+        player0_win = self._check_player_win(self._player_0)
+        player1_win = self._check_player_win(self._player_1)
+        draw = torch.min(game_over, reward == 0)
+
+        return TimeStep(
+            observation=self._boards.clone().detach(),
+            reward=reward.detach(),
+            step_type=step_type.detach(),
+            discount=discount.detach(),
+            prev_action=prev_action.detach(),
+            env_id=self._env_ids,
+            env_info={
+                "play0_win": player0_win.to(torch.float32),
+                "play1_win": player1_win.to(torch.float32),
+                "draw": draw.to(torch.float32),
+                "invalid_move": (~valid).to(torch.float32),
+            })
+
+    def _check_player_win(self, player):
+        B = self._B.unsqueeze(-1).unsqueeze(-1)
+        player = player.unsqueeze(-1).unsqueeze(-1)
+        lines = self._boards[B, self._line_y, self._line_x]
+        return ((lines == player).sum(dim=2) == 3).any(dim=1)
+
+    def _check_game_over(self):
+        board_full = (self._boards == 0).sum(dim=(1, 2)) == 0
+        B = self._B.unsqueeze(-1).unsqueeze(-1)
+        lines = self._boards[B, self._line_y, self._line_x]
+        player1_won = ((lines == -1).sum(dim=2) == 3).any(dim=1)
+        player2_won = ((lines == 1).sum(dim=2) == 3).any(dim=1)
+        return torch.max(board_full, torch.max(player1_won, player2_won))
+
+    def _get_current_player(self):
+        return ((self._boards != 0).sum(dim=(1, 2)) % 2) * 2 - 1
```

Comment:
Here the value of -1 / 1  is also explicitly used. 
Might be able to avoid it if necessary.